kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/listener/CollectingTestEngineListener.kt

```diff
@@ -10,8 +9,8 @@ import kotlin.reflect.KClass
 
 class CollectingTestEngineListener : AbstractTestEngineListener() {
 
-   val specs = concurrentHashMap<KClass<*>, TestResult>()
-   val tests = concurrentHashMap<TestCaseKey, TestResult>()
+   val specs = mutableMapOf<KClass<*>, TestResult>()
+   val tests = mutableMapOf<TestCaseKey, TestResult>()
```

Comment:
Are we worried about race conditions if we just use mutable map ?

Reply:
Seems like we should be. The API with public properties is problematic in this regard, and Kotest doesn't use atomicfu so far, so I'll add a commit introducing a listener-wide Mutex for mutations.

Reply:
Any reason we'd not just use the concurrent hash map ?

Reply:
Yes. The previous `concurrentHashMap` was never thread-safe, but actually returned a `MutableMap`. There is no `ConcurrentHashMap` implementation in stdlib common.

Reply:
Yeah I see. It could use a concurrent hash map on the JVM actual impl though.

Reply:
Indeed that could be an option, since we don't require special `ConcurrentHashMap` methods here. Side effect: Some methods (like `testIgnored()`) mutate multiple structures, which would no longer be atomic then. If results were looked up during ongoing test runs, the maps could be briefly inconsistent (one contains a test, the other does not). What would be the benefit of using expect/actual and `ConcurrentHashMap` instead?

Reply:
This whole class was really just created in order to test kotest itself. It's not meant to be used outside of kotest.
I think ConcurrentHashMap is probably good enough for our purposes?

Reply:
I've tried: `concurrentHashMap()` with 1 "expect" and 3 "actuals" (like formerly but correct on the JVM). Then we're still missing a correct concurrent implementation on Native, and a concurrent implementation for `val names = mutableListOf<String>()`. Lots more code, less correct. Thoughts?

Reply:
Maybe use concurrentHashMap on jvm, and mutableMap on the others, which is better than what we had before?

Reply:
Better than originally, yes. Better than `CollectingTestEngineListener`'s Mutex implementation, which should just work on all platforms? What's wrong with that? What am I missing?

Reply:
Or maybe commit 3281e9c0f826841aae771cae4c0596bf1fc42bea was just overlooked?

Reply:
Yep missed the mutex.
I do wonder if the mutex on the JVM is mapped to something that ensures the JVM MM is adhered to. I assume it is,